tests/gpflow/optimizers/test_scipy.py

```diff
@@ -221,3 +228,276 @@ def f() -> tf.Tensor:
+# entry.
+@pytest.mark.parametrize("compile_cache_size", [1, 2, 3])
+def test_scipy__cache_replacement(compile_cache_size: int) -> None:
+    opt = gpflow.optimizers.Scipy(compile_cache_size=compile_cache_size)
+
+    # Create more closures and variables than the cache can hold
+    variables = [tf.Variable(float(i)) for i in range(compile_cache_size + 1)]
+    closures = [_loss_closure_builder([0], var) for var in variables]
+
+    # Fill the cache
+    for i in range(compile_cache_size):
+        opt.eval_func(closures[i], [variables[i]], tf_fun_args={})
+        assert len(opt.compile_cache) == i + 1
+
+    # Add one more closure to force the cache to replace the oldest entry
+    opt.eval_func(closures[-1], [variables[-1]], tf_fun_args={})
+    assert len(opt.compile_cache) == compile_cache_size
+    assert set(opt.compile_cache.keys()) == set(
+        [
+            (closures[i], (id(variables[i]),), frozenset(), False)
+            for i in range(1, compile_cache_size + 1)
+        ]
+    )
+
+
+# This test checks that the cache is disabled when compile_cache_size is set to 0.
+def test_scipy__cache_disabled() -> None:
+    opt = gpflow.optimizers.Scipy(compile_cache_size=0)
+    variables = _create_variables()
+    counter = [0]
+    closure = _loss_closure_builder(counter, variables)
+
+    # Ensure the cache is empty
+    assert len(opt.compile_cache) == 0
+
+    # Call eval_func multiple times with the same arguments
+    for _ in range(4):
+        opt.minimize(closure, variables, compile=True, tf_fun_args={}, allow_unused_variables=False)
+        # Check that the cache was not used
+        assert len(opt.compile_cache) == 0
+
+    # The closure should be run every time
+    assert counter[0] == 4
+
+
+# This test ensures an error is raised when compile_cache_size is set to a negative value.
+def test_scipy__cache_raises_negative_size() -> None:
+    with pytest.raises(ValueError, match=r"The 'compile_cache_size' argument must be non-negative"):
+        gpflow.optimizers.Scipy(compile_cache_size=-1)
+
+
+# This test ensures that the cache behaves correctly considering all arguments.
+# It verifies that the cache has a miss when any argument changes, and a hit when all arguments
+# remain the same. Note that when `compile` is `False`, the cache is not used and we expect a miss.
+#   - hit : the same closure and variables, same reaming arguments
+#   - miss: the same closure but different variables, same reaming arguments
+#   - miss: Different closure but the same variables, same reaming arguments
+#   - miss: the same closure and variables but different tf_fun_args
+#   - miss: the same closure and variables but different allow_unused_variables
+#   - miss: the same closure, variables and args, but no-compile
+@pytest.mark.parametrize(
+    "expect_cache_hit, same_closure2, same_variables2, same_tf_fun_args2, "
+    "allow_unused_variables2, compile",
+    [
+        (True, True, True, True, False, True),
```

Comment:
Maybe use `pytest.param` to give the test cases human-readable names so it's clearer what each one does in case one fails? (you could probably just move the comment descriptions above into the id)